Shell/Shell.cpp

```diff
@@ -1359,6 +1360,69 @@ void Shell::custom_event(Core::CustomEvent& event)
             Core::EventLoop::current().post_event(*this, make<Core::CustomEvent>(ShellEventType::ReadLine));
         return;
     }
+
+    if (event.custom_type() == ChildEvent) {
+#ifdef SH_DEBUG
+        dbg() << "SIGCHLD!";
+#endif
+        auto& jobs = this->jobs;
+        Vector<u64> disowned_jobs;
+        // Workaround the fact that we can't receive *who* exactly changed state.
+        // The child might still be alive (and even running) when this signal is dispatched to us
+        // so just...repeat until we find a suitable child.
+        // This, of course, will mean that someone can send us a SIGCHILD and we'd be spinning here forever.
+        bool found_child = false;
+        for (auto& it : jobs) {
+            auto job_id = it.key;
+            auto& job = *it.value;
+            int wstatus = 0;
+#ifdef SH_DEBUG
+            dbgf("waitpid({}) = ...", job.pid());
+#endif
+            auto child_pid = waitpid(job.pid(), &wstatus, WNOHANG | WUNTRACED);
```

Comment:
If you're waiting for any process in a process group, shouldn't it be `-job.pid()`?

Reply:
We're waiting for a specific process here, which should always be the process group leader.
Not sure why we'd want to look at the other members of a process group, am I missing something?

Reply:
No I guess you're right. Sorry I'm not really familiar with all that ;-)